### PR TITLE
[1.24] pkg/controller/job: re-honor exponential backoff

### PR DESCRIPTION
**Pull request by nikhita**
Cherry pick of #114516 on release-1.24.

#114516: pkg/controller/job: re-honor exponential backoff delay

For details on the cherry pick process, see the [cherry pick requests](https://git.k8s.io/community/contributors/devel/sig-release/cherry-picks.md) page.

```release-note
Failed pods associated with a job with `parallelism = 1` are recreated by the job controller honoring exponential backoff delay again. However, for jobs with `parallelism > 1`, pods might be created without exponential backoff delay.
```

### CODE REVIEW

**Comment by Verolop** [comment]
hi @nikhita @soltysh , we need a lgtm & approval to be able to include this (and the other related cherry-picks) in the January patches. Thanks!

**Review by soltysh — APPROVED** [review]
/lgtm
/approve
/triage accepted

**Comment by cpanato** [comment]
cherry pick deadline is over this will get merged after the patches this week

/hold

**Review by xmudrii — APPROVED** [review]
As discussed on Slack, this is a serious regression and we want to get it fixed. Ref: https://kubernetes.slack.com/archives/CJH2GBF7Y/p1673881966562209

/lgtm
/approve

**Pull request merged**
